Providing background for Baremetal IPI based enhancements (#102)
openshift/enhancements: enhancements/baremetal-IPI/baremetal.md

Installer is responsble for providing all the necessary configs required for
a successful deployment.

## Motivation

The motivation for this enhancement request is to provide a background for all the
the subsequent enhancement requests for Baremetal IPI deployments.

### Goals

The goal of this enhancement request is to provide context for all the changes
that have already been merged towards making Baremetal IPI deployments a reality.
All future Baremetal enhancement requests will refer back to this one to provide
context.

### Non-Goals

Raising development PRs as a result of this enhancement request.

## Proposal

Every OpenShift based Baremetal IPI deployment will run a "metal3" pod on
one Master Node. A "metal3" pod includes a container running BareMetal
Operator(BMO) and several other supporting containers that work together.

The BMO and other supporting containers together are able to discover a
baremetal server in a pre-determined provisioning network, learn the
HW attributes of the server and eventually boot it to make it available
as a Machine within a MachineSet.

The Machine API Operator (MAO) currently deploys the "metal3" pod only

> when the Platform type is `BareMetal` but installs the BaremetalHost CRD

Reviewer: just to clarify the BaremetalHost CRD is exposed by the mao as part of the release payload which is managed by the cluster version operator.

Author: Fixed.

for all platform types starting from release 4.2.

The installer is responsible for kicking off a Baremetal IPI deployment
with the right configuration.

A set of follow-up enhancement requests will try to bring in the following
enhancements:

> 1. Providing the BMO and it supporting containers with configuration that

Reviewer: typo it/its

Reviewer: I'd probably just keep this to represent current state of the world so there's no confusion this enhancements are discussed separately

Author: Agreed.

is specific to the server and its provisioning environment. This enhancement
request is already under review [2]. 

2. The MAO would also be responsible for deploying a DeamonSet that would
run on all baremetal servers to monitor the hardware inventory on each
of these servers.

### User Stories

With the addition of features described in this and other enhancements
detailed in this current directory, OpenShift can be used to bring up
a functioning cluster starting with a set of baremetal servers. As
mentioned earlier, these enhancements rely on the Baremetal Operator (BMO)
[1] running within the "metal3" pod to manage baremetal hosts. The BMO in

> turn relies on the Ironic service [3] to manage and provision baremetal

Reviewer: would be nice to format the user stories as "As a role I want to feature so I... "

---

> title: Neat-Enhancement-Idea

Reviewer 2: It is a neat idea, but I think you want the actual doc title here? ;-)

Author: If you say so, @dhellmann :-). Needed to check if it had to be a hyphenated string and didn't get back to it.

authors:
  - "@sadasu"
reviewers:
  - TBD
  - "@abhinavdahiya"
approvers:
  - TBD
  - "@abhinavdahiya"
creation-date: 2019-11-06
last-updated: 2019-11-06
status: implemented
---

# Adding Baremetal IPI capabilities to OpenShift

This enhancement serves to provide the context for a whole slew of features

> and enhancements that will follow to make Baremetal IPI deployments via

Reviewer 3: Can we add a definition for `IPI` at the top here?

Author: Done.

OpenShift a reality.

At the time of this writing, code for some of the enhancements have already
merged, some are in progress and others are yet to me implemented. References
to all these features in different stages of development will be provided
below.

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift/docs]

## Open Questions [optional]

> This is where to call out areas of the design that require closure before deciding

Reviewer 2: Is this boilerplate?

Author: Yes. Kept it there for now as guidance if reviewers feel I should be adding something to a section.

to implement the design.  For instance, 
 > 1. This requires exposing previously private resources which contain sensitive
  information.  Can we do this? 

## Summary

Baremetal IPI deployments enable OpenShift to enroll baremetal servers to become

> Machines that can be consumed to run K8s workloads.

Reviewer 2: ```suggestion
Nodes that can run K8s workloads.
```

Author: Done.

The Baremetal Operator (https://github.com/metal3-io/baremetal-operator) along
with other provisioning services (Ironic and dependencies) run in their own

> pod called the "metal3". This pod is deployed by the Machine API Operator when

Reviewer 2: ```suggestion
pod called "metal3". This pod is deployed by the Machine API Operator when
```

Author: Done

> the Platform type is Baremetal. The OpenShift Installer is responsble for

Reviewer 2: We should use the literal value here. Is it "Baremetal" with a capital B? If so, maybe wrap that word in back-quotes so it shows up as a literal?

Author: Done.

providing all the necessary configs required for a successful deployment.

## Motivation

The motivation for this enhancement request is to provide a background for all the
the subsequent enhancement requests for Baremetal IPI deployments.

### Goals

The goal of this enhancement request is to provide context for all the changes
that have already been merged towards making Baremetal IPI deployments a reality.
All future Baremetal enhancement requests will refer back to this one to provide
context.

### Non-Goals

> No development PRs will be raised as a result of this enhancement request.

Reviewer 4: This may be a double negative. The non-goal (IE: what will NOT be done) is no development prs [....].

Author: Fixed it.

## Proposal

Every OpenShift based Baremetal IPI deployment will run a "metal3" pod on

> all Master Nodes. A "metal3" pod includes a container running BareMetal

Reviewer 3: Is the `all Master Nodes` part important or an implementation detail to ensure high availability for the operator?

Reviewer 5: it only has one replica actually.. so its run on one of the master nodes.

Author: Thanks @imain .

Operator(BMO) and several other supporting containers that work togther.

The BMO and other supporting containers together are able to discover a
baremetal server in a pre-determined provisioning network, learn the
HW attributes of the server and eventually boot it to make it available
as a Machine within a MachineSet.

The Machine API Operator currently deploys the "metal3" pod only when the

> Platform type is Baremetal. The MAO also installs the BaremetalHostCRD

Reviewer 2: ```suggestion
Platform type is Baremetal. The MAO also installs the BaremetalHost CRD
```

> Platform type is Baremetal. The MAO also installs the BaremetalHostCRD

Reviewer 5: I know we're used to the MAO acronym but we should probably expand it here..

Author: Added the acronym with the 1st full mention of the Machine API Operator earlier in the doc.

for every type of deployment starting from release 4.2.

The installer is responsible for kicking off a Baremetal IPI deployment
with the right configuration.

A set of follow-up enhancement requests will try to bring in the following
enhancements:
1. Providing the BMO and it supporting containers with configuration that
is specific to the server and its provisioning environment. This enhancement
request is already under review here : 
https://github.com/openshift/enhancements/pull/90

2. The MAO would also be responsible for deploying a DeamonSet that would
run on all baremetal servers to monitor the hardware inventory on each
of these servers.

> ### User Stories [optional]

Reviewer 2: We should either remove this section, or remove the "(optional)" and add some stories like deploying the control plane, adding workers, and removing workers.

Reviewer 5: I wonder id we shouldn't describe how a new node gets provisioned from start to finish in here somewhere to really explain the idea in full?

Author: The latest version has this information. Please let me know if everything has been captured accurately.

Detail the things that people will be able to do if this is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.

#### Story 1

#### Story 2

### Implementation Details/Notes/Constraints [optional]

The MAO already deploys the "metal3" pods with the BareMetal Operator and the
other (Ironic) containers. Currently, there is no mechanism in place to provide
these containers with their configuration.

### Risks and Mitigations

Will be specified in follow-up enhancement requests mentioned above.

## Design Details

### Test Plan

True e2e and integration testing can happen only after implementation for
enhancement https://github.com/openshift/enhancements/pull/90 lands.
Until then, e2e testing is being performed with the help of some
developer scripts.

Unit tests have been added to MAO and the Installer to test additions
made for the Baremetal IPI case.

> ### Graduation Criteria

Reviewer 2: Maybe in this section we should say that the metal3 integration is already Tech Preview for 4.2 and scheduled for GA in whatever release we've promised it will be.

**Note:** *Section not required until targeted at a release.*

Define graduation milestones.

These may be defined in terms of API maturity, or as something else. Initial proposal
should keep this high-level with a focus on what signals will be looked at to
determine graduation.

Consider the following in developing the graduation criteria for this
enhancement:
- Maturity levels - `Dev Preview`, `Tech Preview`, `GA`
- Deprecation

Clearly define what graduation means.

#### Examples

These are generalized examples to consider, in addition to the aforementioned
[maturity levels][maturity-levels].

##### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation, relative API stability
- Sufficient test coverage
- Gather feedback from users rather than just developers

##### Tech Preview -> GA 

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default

**For non-optional features moving to GA, the graduation criteria must include
end to end tests.**

> ##### Removing a deprecated feature

Reviewer 2: I think we can remove this section.

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

> ### Upgrade / Downgrade Strategy

Reviewer 2: The MAO handles upgrades for us, right? We haven't had any API changes to deal with, yet.

If applicable, how will the component be upgraded and downgraded? Make sure this
is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade in order to keep previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade in order to make use of the enhancement?

> ### Version Skew Strategy

Reviewer 2: We're locked to the cluster version because metal3 is part of the release image. I'm not sure if that means this doesn't apply?

How will the component handle version skew with other components?
What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- During an upgrade, we will always have skew among components, how will this impact your work?
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI, CRI
  or CNI may require updating that component before the kubelet.

> ## Implementation History

Reviewer 2: We should definitely mention the current status in 4.2 here.

Major milestones in the life cycle of a proposal should be tracked in `Implementation
History`.

## Drawbacks

> The idea is to find the best form of an argument why this enhancement should _not_ be implemented.

Reviewer 2: Users should only ever use clouds? :-)

## Alternatives

> Similar to the `Drawbacks` section the `Alternatives` section is used to

Reviewer 2: One alternative would be to deliver metal3 on top of the installed cluster, except that doesn't help us install the cluster in the first place. That's why the integration with the installer is there.

Reviewer 3: The installer tries to address this sort of thing today with its bootstrap node:
> OpenShift breaks this dependency loop using a temporary bootstrap machine. This bootstrap machine is booted with a concrete Ignition Config which describes how to create the cluster. This machine acts as a temporary control plane whose sole purpose is launching the rest of the cluster.

ref: https://github.com/openshift/installer/blob/master/docs/user/overview.md#cluster-installation-process

The full bootstrap model hasn't been fully actualized: he installer still creates the master node resources rather than bootstrapping through the cluster-api like workers are. It might be worth looking into helping that effort along if it would solve some of your use cases.

Reviewer 6: Baremetal IPI is already integrated with the installer and the bootstrap node/terraform:

https://github.com/openshift/installer/blob/master/docs/user/metal/install_ipi.md

But yes it would be interesting if we could help move away from terraform and towards using the metal3 operator on the bootstrap node.

highlight and record other possible approaches to delivering the value proposed
by an enhancement.

> ## Infrastructure Needed [optional]

Reviewer 2: This section could link to the metal3-io and openshift-metal3 orgs on GitHub. Everything exists there and the pieces we need have been imported into the openshift org already, so there's no work to do.

Author: Yes, added some information about this.
